File: code/jvm/repository/src/testFixtures/java/pt/isel/repositories/mem/ChannelMemberRepositoryInMem.java
package pt.isel.repositories.mem;

import pt.isel.domain.channels.AccessType;
import pt.isel.domain.channels.Channel;
import pt.isel.domain.channels.ChannelMember;
import pt.isel.domain.users.UserInfo;
import pt.isel.repositories.channels.ChannelMemberRepository;

import java.util.ArrayList;
import java.util.List;

public class ChannelMemberRepositoryInMem implements ChannelMemberRepository {
    private final List<ChannelMember> members = new ArrayList<>();
    private long nextId = 1;

    @Override
    public ChannelMember addUserToChannel(UserInfo userInfo, Channel channel, AccessType accessType) {
        if (findUserInChannel(userInfo.id(), channel.id()) != null)
            throw new RuntimeException("channel_members_user_id_channel_id_key");
        ChannelMember member = new ChannelMember(nextId++, userInfo, channel, accessType);
        members.add(member);
        return member;
    }

    @Override
    public ChannelMember findUserInChannel(Long userId, Long channelId) {
        return members.stream()
                .filter(m -> m.user().id().equals(userId) && m.channel().id().equals(channelId))
                .findFirst()
                .orElse(null);
    }

    @Override
    public List<ChannelMember> findAllChannelsForUser(Long userId, int limit, int offset) {
        return members.stream()
                .filter(m -> m.user().id().equals(userId))
                .skip(offset)
                .limit(limit)
                .toList();
    }

    @Override
    public List<ChannelMember> findAllMembersInChannel(Long channelId, int limit, int offset) {
        return members.stream()
                .filter(m -> m.channel().id().equals(channelId))
                .skip(offset)
                .limit(limit)
                .toList();
    }

    @Override
    public void removeUserFromChannel(Long userId, Long channelId) {
        members.removeIf(m -> m.user().id().equals(userId) && m.channel().id().equals(channelId));
    }

    @Override
    public ChannelMember findById(Long id) {
        return members.stream().filter(m -> m.id().equals(id)).findFirst().orElse(null);
    }

    @Override
    public List<ChannelMember> findAll() {
        return new ArrayList<>(members);
    }

    @Override
    public void save(ChannelMember entity) {
        members.removeIf(m -> m.id().equals(entity.id()));
        members.add(entity);
    }

    @Override
    public void deleteById(Long id) {
        members.removeIf(m -> m.id().equals(id));
    }

    @Override
    public void clear() {
        members.clear();
        nextId = 1;
    }
}

File: code/jvm/repository/src/test/kotlin/pt/isel/repositories/mem/ChannelMemberRepositoryInMem.kt
package pt.isel.repositories.mem

import jakarta.inject.Named
import pt.isel.domain.channels.AccessType
import pt.isel.domain.channels.Channel
import pt.isel.domain.channels.ChannelMember
import pt.isel.domain.users.UserInfo
import pt.isel.repositories.channels.ChannelMemberRepository

/**
 * Naif in memory repository non thread-safe and basic sequential id. Useful for unit tests purpose.
 */
@Named
class ChannelMemberRepositoryInMem : ChannelMemberRepository {
    private val channelMembers = mutableListOf<ChannelMember>()

    override fun addUserToChannel(
        userInfo: UserInfo,
        channel: Channel,
        accessType: AccessType,
    ): ChannelMember =
        ChannelMember(
            channelMembers.size.toLong() + 1,
            userInfo,
            channel,
            accessType,
        ).also {
            channelMembers.add(it)
        }

    override fun findById(id: Long): ChannelMember? = channelMembers.firstOrNull { it.id == id }

    override fun findUserInChannel(
        userId: Long,
        channelId: Long,
    ): ChannelMember? = channelMembers.firstOrNull { it.user.id == userId && it.channel.id == channelId }

    override fun findAllChannelsForUser(
        userId: Long,
        limit: Int,
        offset: Int,
    ): List<ChannelMember> = channelMembers.filter { it.user.id == userId }.drop(offset).take(limit)

    override fun findAllMembersInChannel(
        channelId: Long,
        limit: Int,
        offset: Int,
    ): List<ChannelMember> = channelMembers.filter { it.channel.id == channelId }.drop(offset).take(limit)

    override fun findAll(): List<ChannelMember> = channelMembers.toList()

    override fun save(entity: ChannelMember) {
        channelMembers.removeIf { it.id == entity.id }.apply { channelMembers.add(entity) }
    }

    override fun removeUserFromChannel(
        userId: Long,
        channelId: Long,
    ) {
        channelMembers.removeIf { it.user.id == userId && it.channel.id == channelId }
    }

    override fun deleteById(id: Long) {
        channelMembers.removeIf { it.id == id }
    }

    override fun clear() {
        channelMembers.clear()
    }
}
